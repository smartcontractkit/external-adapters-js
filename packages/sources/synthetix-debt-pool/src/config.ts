import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as DefaultConfig } from '@chainlink/types'

export const NAME = 'SYNTHETIX_DEBT_POOL'

export const DEFAULT_ENDPOINT = 'debt'

export enum SUPPORTED_CHAINS {
  ETHEREUM = 'ETHEREUM',
  OPTIMISM = 'OPTIMISM',
}

export interface Config extends DefaultConfig {
  chains: {
    [key: string]: {
      rpcUrl: string
      addressProviderContractAddress: string
    }
  }
}

export const makeConfig = (prefix?: string): Config => {
  const config: Config = {
    ...Requester.getDefaultConfig(prefix),
    defaultEndpoint: DEFAULT_ENDPOINT,
    chains: {},
  }

  for (const chainName of Object.values(SUPPORTED_CHAINS)) {
    const chainRpcURL = util.getEnv(`${chainName}_RPC_URL`, prefix)
    const addressProviderContractAddress = util.getEnv(
      `${chainName}_ADDRESS_PROVIDER_CONTRACT_ADDRESS`,
      prefix,
    )
    if (chainRpcURL && addressProviderContractAddress) {
      config.chains[chainName] = {
        rpcUrl: chainRpcURL,
        addressProviderContractAddress,
      }
    }
  }

  const chains = Object.keys(config.chains)
  if (chains.length === 0) throw Error('Must set at least one RPC Chain URL')
  return config
}
