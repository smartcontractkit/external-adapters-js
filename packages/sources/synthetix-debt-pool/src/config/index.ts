import { AdapterConfigError, Requester, util } from '@chainlink/ea-bootstrap'
import { Config as DefaultConfig } from '@chainlink/ea-bootstrap'

export const NAME = 'SYNTHETIX_DEBT_POOL'

export const DEFAULT_ENDPOINT = 'debt'

export const ENV_RPC_URL = 'RPC_URL'
export const ENV_CHAIN_ID = 'CHAIN_ID'

export const DEFAULT_ETHEREUM_CHAIN_ID = '1'
export const DEFAULT_KOVAN_CHAIN_ID = '42'
export const DEFAULT_OPTIMISM_CHAIN_ID = '10'
export const DEFAULT_KOVAN_OPTIMISM_CHAIN_ID = '69'
export const DEFAULT_GOERLI_CHAIN_ID = '5'
export const DEFAULT_GOERLI_OPTIMISM_CHAIN_ID = '420'
export const DEFAULT_SEPOLIA_CHAIN_ID = '11155111'
export const DEFAULT_SEPOLIA_OPTIMISM_CHAIN_ID = '11155420'

export enum SupportedChains {
  ETHEREUM = 'mainnet',
  OPTIMISM = 'mainnet-ovm',
  KOVAN = 'kovan',
  KOVAN_OPTIMISM = 'kovan-ovm',
  GOERLI = 'goerli',
  GOERLI_OPTIMISM = 'goerli-ovm',
  SEPOLIA = 'sepolia',
  SEPOLIA_OPTIMISM = 'sepolia-ovm',
}

export interface Config extends DefaultConfig {
  chains: {
    [key: string]: {
      rpcURL: string
      chainId: string | number | undefined
      chainAddressResolverProxyAddress: string
    }
  }
}

export const makeConfig = (prefix?: string): Config => {
  const config: Config = {
    ...Requester.getDefaultConfig(prefix),
    defaultEndpoint: DEFAULT_ENDPOINT,
    chains: {},
  }

  for (const chainName of Object.values(SupportedChains)) {
    const envVarPrefix = getRPCUrlPrefix(chainName)
    const chainRpcURL = util.getEnv(ENV_RPC_URL, envVarPrefix)
    const chainId =
      parseInt(util.getEnv(ENV_CHAIN_ID, envVarPrefix) || getDefaultChainId(chainName)) ||
      util.getEnv(ENV_CHAIN_ID, envVarPrefix)

    const chainAddressResolverProxyAddress =
      util.getEnv('ADDRESS_RESOLVER_PROXY_CONTRACT_ADDRESS', envVarPrefix) ||
      getDefaultAddressResolverProxyAddress(chainName)
    if (chainRpcURL) {
      config.chains[chainName] = {
        rpcURL: chainRpcURL,
        chainId,
        chainAddressResolverProxyAddress,
      }
    }
  }

  const chains = Object.keys(config.chains)
  if (chains.length === 0)
    throw new AdapterConfigError({ message: 'Must set at least one RPC Chain URL' })
  return config
}

const getRPCUrlPrefix = (networkName: SupportedChains): string => {
  switch (networkName) {
    case SupportedChains.ETHEREUM:
      return ''
    case SupportedChains.KOVAN:
      return 'KOVAN'
    case SupportedChains.OPTIMISM:
      return 'OPTIMISM'
    case SupportedChains.KOVAN_OPTIMISM:
      return 'KOVAN_OPTIMISM'
    case SupportedChains.GOERLI:
      return 'GOERLI'
    case SupportedChains.GOERLI_OPTIMISM:
      return 'GOERLI_OPTIMISM'
    case SupportedChains.SEPOLIA:
      return 'SEPOLIA'
    case SupportedChains.SEPOLIA_OPTIMISM:
      return 'SEPOLIA_OPTIMISM'
  }
}

const getDefaultAddressResolverProxyAddress = (networkName: SupportedChains): string => {
  switch (networkName) {
    case SupportedChains.ETHEREUM:
      return '0x4E3b31eB0E5CB73641EE1E65E7dCEFe520bA3ef2'
    case SupportedChains.KOVAN:
      return '0x242a3DF52c375bEe81b1c668741D7c63aF68FDD2'
    case SupportedChains.OPTIMISM:
      return '0x1Cb059b7e74fD21665968C908806143E744D5F30'
    case SupportedChains.KOVAN_OPTIMISM:
      return '0x7a6f9eDDC03Db81927eA4131919343f93CA9b6a7'
    case SupportedChains.GOERLI:
      return '0x58719E8Ef4d201541e44505a2ACB3424481d6681'
    case SupportedChains.GOERLI_OPTIMISM:
      return '0x9Fc84992dF5496797784374B810E04238728743d'
    case SupportedChains.SEPOLIA:
      return '0xb61a6aF69e992e1bed69b0aE0CBa5143CA25D4D1'
    case SupportedChains.SEPOLIA_OPTIMISM:
      return '0x529C553eF2d0370279DC8AbF19702B98b166D252'
  }
}

const getDefaultChainId = (networkName: SupportedChains): string => {
  switch (networkName) {
    case SupportedChains.ETHEREUM:
      return DEFAULT_ETHEREUM_CHAIN_ID
    case SupportedChains.KOVAN:
      return DEFAULT_KOVAN_CHAIN_ID
    case SupportedChains.OPTIMISM:
      return DEFAULT_OPTIMISM_CHAIN_ID
    case SupportedChains.KOVAN_OPTIMISM:
      return DEFAULT_KOVAN_OPTIMISM_CHAIN_ID
    case SupportedChains.GOERLI:
      return DEFAULT_GOERLI_CHAIN_ID
    case SupportedChains.GOERLI_OPTIMISM:
      return DEFAULT_GOERLI_OPTIMISM_CHAIN_ID
    case SupportedChains.SEPOLIA:
      return DEFAULT_SEPOLIA_CHAIN_ID
    case SupportedChains.SEPOLIA_OPTIMISM:
      return DEFAULT_SEPOLIA_OPTIMISM_CHAIN_ID
  }
}
