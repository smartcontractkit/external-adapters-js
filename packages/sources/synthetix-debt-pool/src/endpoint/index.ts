export * as debt from './debt'
