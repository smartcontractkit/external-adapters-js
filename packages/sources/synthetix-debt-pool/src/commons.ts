import { Validator, Requester, Logger } from '@chainlink/ea-bootstrap'
import { AdapterRequest, AdapterResponse, InputParameters } from '@chainlink/ea-bootstrap'
import { BigNumber, ethers } from 'ethers'
import { SupportedChains, Config } from './config'
import { AdapterError } from '@chainlink/ea-bootstrap'
import { SynthetixJS, synthetix } from '@synthetixio/contracts-interface'

export type TInputParameters = { chainSources: string[] }
export const inputParameters: InputParameters<TInputParameters> = {
  chainSources: {
    required: false,
    description: `Array of chains to pull debt from. Options for array elements are "mainnet", "mainnet-ovm", "kovan", "kovan-ovm"`,
    type: 'array',
  },
}

type GetDebtData = (jobRunID: string, config: Config, chainsToQuery: string[]) => Promise<BigNumber>

export const getDataFromAcrossChains = async (
  request: AdapterRequest,
  config: Config,
  getDebtData: GetDebtData,
): Promise<AdapterResponse> => {
  const validator = new Validator<TInputParameters>(request, inputParameters)
  const jobRunID = validator.validated.id
  let { chainSources } = validator.validated.data

  if (!chainSources) {
    Logger.info('chainSources is undefined.  Will aggregate debt over all supported chains')
    chainSources = [SupportedChains.ETHEREUM, SupportedChains.OPTIMISM]
  }
  validateChainSources(jobRunID, chainSources)
  const debt = await getDebtData(jobRunID, config, chainSources)
  const result = {
    data: {
      result: debt.toString(),
    },
  }
  return Requester.success(jobRunID, result, config.verbose)
}

const validateChainSources = (jobRunID: string, chainSources: string[]) => {
  const supportedChains = Object.values(SupportedChains) as string[]
  for (const source of chainSources) {
    if (!supportedChains.includes(source)) {
      throw new AdapterError({
        jobRunID,
        message: `${source} is not a supported chain.  Must be one of ${supportedChains.join(',')}`,
      })
    }
  }
}

export const getChainSynthetixInstance = (
  network: string,
  jobRunID: string,
  config: Config,
): SynthetixJS => {
  const { rpcURL, networkId } = config.chains[network]
  if (!rpcURL) {
    throw new AdapterError({
      jobRunID,
      message: `RPC URL not set for chain: ${network}`,
    })
  }
  const provider = new ethers.providers.JsonRpcProvider(rpcURL)
  return synthetix({ provider, networkId })
}
