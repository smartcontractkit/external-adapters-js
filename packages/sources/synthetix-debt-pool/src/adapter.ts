import { AdapterRequest, ExecuteWithConfig, APIEndpoint, ExecuteFactory } from '@chainlink/types'
import { makeConfig, Config as AdapterConfig } from './config'
import * as endpoints from './endpoint'
import { Builder } from '@chainlink/ea-bootstrap'

export const execute: ExecuteWithConfig<AdapterConfig> = async (request, context, config) => {
  return Builder.buildSelector(request, context, config, endpoints)
}

export const endpointSelector = (request: AdapterRequest): APIEndpoint<AdapterConfig> =>
  Builder.selectEndpoint(request, makeConfig(), endpoints)

export const makeExecute: ExecuteFactory<AdapterConfig> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
