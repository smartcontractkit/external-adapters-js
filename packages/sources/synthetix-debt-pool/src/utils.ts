import {
  Validator,
  Requester,
  Logger,
  AdapterInputError,
  AdapterDataProviderError,
  AdapterConnectionError,
  AdapterConfigError,
} from '@chainlink/ea-bootstrap'
import { AdapterRequest, AdapterResponse, InputParameters } from '@chainlink/ea-bootstrap'
import { ethers } from 'ethers'
import { SupportedChains, Config } from './config'
import { AdapterError } from '@chainlink/ea-bootstrap'
import { READ_PROXY_ABI, ADDRESS_RESOLVER_ABI } from './endpoint/abi'

export type TInputParameters = { chainSources: string[] }
export const inputParameters: InputParameters<TInputParameters> = {
  chainSources: {
    required: false,
    description: `Array of chains to pull debt from. Options for array elements are "mainnet", "mainnet-ovm", "kovan", "kovan-ovm", "goerli", "goerli-ovm", "sepolia", "sepolia-ovm"`,
    type: 'array',
  },
}

type GetDebtData = (
  jobRunID: string,
  config: Config,
  chainsToQuery: [string, number][],
) => Promise<ethers.BigNumber>

export const getDataFromAcrossChains = async (
  request: AdapterRequest,
  config: Config,
  getDebtData: GetDebtData,
): Promise<AdapterResponse> => {
  const validator = new Validator(request, inputParameters)
  const jobRunID = validator.validated.id
  let { chainSources } = validator.validated.data

  if (!chainSources) {
    Logger.info('chainSources is undefined.  Will aggregate debt over all supported chains')
    chainSources = [SupportedChains.ETHEREUM, SupportedChains.OPTIMISM]
  }
  validateChainSources(jobRunID, chainSources)
  // Pin to current moment's latest blocks to ensure time consistency
  const latestBlockByChain = await getLatestBlockByChain(jobRunID, config, chainSources)
  const debt = await getDebtData(jobRunID, config, latestBlockByChain)
  const result = {
    data: {
      result: debt.toString(),
    },
  }
  return Requester.success(jobRunID, result, config.verbose)
}

const validateChainSources = (jobRunID: string, chainSources: string[]) => {
  const supportedChains = Object.values(SupportedChains) as string[]
  for (const source of chainSources) {
    if (!supportedChains.includes(source)) {
      throw new AdapterInputError({
        jobRunID,
        message: `${source} is not a supported chain.  Must be one of ${supportedChains.join(',')}`,
      })
    }
  }
}

export const getAddressResolver = async (
  provider: ethers.providers.JsonRpcProvider,
  addressResolverProxyAddress: string,
  jobRunID: string,
  network: string,
): Promise<string> => {
  try {
    const addressResolver = new ethers.Contract(
      addressResolverProxyAddress,
      READ_PROXY_ABI,
      provider,
    )
    return await addressResolver.target()
  } catch (e) {
    return errorResponse(
      e,
      jobRunID,
      network,
      `Failed to fetch address resolver. Error Message: ${e}`,
    )
  }
}

export const getContractAddress = async (
  provider: ethers.providers.JsonRpcProvider,
  addressResolverAddress: string,
  contractName: string,
  jobRunID: string,
  network: string,
): Promise<string> => {
  try {
    const addressResolver = new ethers.Contract(
      addressResolverAddress,
      ADDRESS_RESOLVER_ABI,
      provider,
    )
    const contractNameBytes32 = ethers.utils.formatBytes32String(contractName)
    return await addressResolver.getAddress(contractNameBytes32)
  } catch (e) {
    return errorResponse(
      e,
      jobRunID,
      network,
      `Failed to fetch ${contractName} contract address from resolver address - ${addressResolverAddress}. Error Message: ${e}`,
    )
  }
}

export const getSynthetixBridgeName = (networkName: string, jobRunID: string): string => {
  if (
    networkName === SupportedChains.ETHEREUM ||
    networkName === SupportedChains.KOVAN ||
    networkName === SupportedChains.GOERLI ||
    networkName === SupportedChains.SEPOLIA
  )
    return 'SynthetixBridgeToOptimism'
  if (
    networkName === SupportedChains.OPTIMISM ||
    networkName === SupportedChains.KOVAN_OPTIMISM ||
    networkName === SupportedChains.GOERLI_OPTIMISM ||
    networkName === SupportedChains.SEPOLIA_OPTIMISM
  )
    return 'SynthetixBridgeToBase'
  throw new AdapterInputError({
    jobRunID,
    message: `${networkName} is not a supported network.}`,
  })
}

export const getDebtMigratorName = (networkName: string, jobRunID: string): string => {
  if (
    networkName === SupportedChains.ETHEREUM ||
    networkName === SupportedChains.GOERLI ||
    networkName === SupportedChains.SEPOLIA
  )
    return 'DebtMigratorOnEthereum'
  if (
    networkName === SupportedChains.OPTIMISM ||
    networkName === SupportedChains.GOERLI_OPTIMISM ||
    networkName === SupportedChains.SEPOLIA_OPTIMISM
  )
    return 'DebtMigratorOnOptimism'
  throw new AdapterInputError({
    jobRunID,
    message: `${networkName} is not a supported network.}`,
  })
}

export const getLatestBlockByChain = async (
  jobRunID: string,
  config: Config,
  chainsToQuery: string[],
): Promise<[string, number][]> =>
  await Promise.all(
    chainsToQuery.map(async (network): Promise<[string, number]> => {
      if (!config.chains[network])
        throw new AdapterConfigError({
          jobRunID,
          statusCode: 500,
          message: `Chain ${network} not configured`,
        })

      try {
        const networkProvider = new ethers.providers.JsonRpcProvider(config.chains[network].rpcURL)
        const latestBlock = await networkProvider.getBlockNumber()
        return [network, latestBlock]
      } catch (e: any) {
        const error = e as any
        const errorPayload = {
          jobRunID,
          message: `Failed to fetch latest block data from chain ${network}.  Error Message: ${error}`,
        }
        throw error.response
          ? new AdapterDataProviderError(errorPayload)
          : error.request
          ? new AdapterConnectionError(errorPayload)
          : new AdapterError(errorPayload)
      }
    }),
  )

export const errorResponse = (error: any, jobRunID: string, network: string, message: string) => {
  const errorPayload = {
    jobRunID,
    network,
    message,
  }
  throw error.response
    ? new AdapterDataProviderError(errorPayload)
    : error.request
    ? new AdapterConnectionError(errorPayload)
    : new AdapterError(errorPayload)
}
