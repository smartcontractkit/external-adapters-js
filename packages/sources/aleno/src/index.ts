import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
