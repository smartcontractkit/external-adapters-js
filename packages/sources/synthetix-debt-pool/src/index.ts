import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig, NAME } from './config'

const server = expose(NAME, makeExecute()).server

export { NAME, makeExecute, makeConfig, server }
