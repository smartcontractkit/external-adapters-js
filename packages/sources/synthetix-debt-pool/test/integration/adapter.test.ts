import { server as startServer } from '../../src/index'
import { BigNumber, utils } from 'ethers'
import request, { SuperTest, Test } from 'supertest'
import http from 'http'
import process from 'process'
import { AddressInfo } from 'net'

const mockChainConfig = {
  ethereum: {
    rpcUrl: 'fake-ethereum-rpc-url',
    addressProviderContractAddress: 'fake-ethereum-address-provider',
    debtCacheAddress: 'fake-ethereum-debt-cache-address',
    synthetixDebtShareAddress: 'fake-ethereum-synthetix-debt-share-address',
  },
  optimism: {
    rpcUrl: 'fake-optimism-rpc-url',
    addressProviderContractAddress: 'fake-optimism-address-provider',
    debtCacheAddress: 'fake-optimism-debt-cache-address',
    synthetixDebtShareAddress: 'fake-optimism-synthetix-debt-share-address',
  },
}

const getContractAddress = (contractNameBytes: string, chain: string): string => {
  switch (contractNameBytes) {
    case utils.formatBytes32String('DebtCache'):
      return mockChainConfig[chain].debtCacheAddress
    case utils.formatBytes32String('SynthetixDebtShare'):
      return mockChainConfig[chain].synthetixDebtShareAddress
  }
}

const mockEthereumAddressProviderContract = {
  getAddress: (contractName: string) => getContractAddress(contractName, 'ethereum'),
}

const mockOptimismAddressProviderContract = {
  getAddress: (contractName: string) => getContractAddress(contractName, 'optimism'),
}

const mockEthereumDebtCacheContract = {
  currentDebt: () => [BigNumber.from('274504021465419663278269593'), false],
  totalNonSnxBackedDebt: () => [BigNumber.from('388546283057244275166159'), false],
}

const mockOptimismDebtCacheContract = {
  currentDebt: () => [BigNumber.from('50977793699622560436740360'), false],
  totalNonSnxBackedDebt: () => [BigNumber.from('18881943681246986146020'), false],
}

const mockSynthetixDebtShareContract = {
  totalSupply: () => BigNumber.from('214522823281993900095205964'),
}

jest.mock('ethers', () => ({
  ...jest.requireActual('ethers'),
  ethers: {
    providers: {
      JsonRpcProvider: function (_: string) {
        return {}
      },
    },
    Contract: function (address: string) {
      switch (address) {
        case mockChainConfig.ethereum.addressProviderContractAddress:
          return mockEthereumAddressProviderContract
        case mockChainConfig.optimism.addressProviderContractAddress:
          return mockOptimismAddressProviderContract
        case mockChainConfig.ethereum.debtCacheAddress:
          return mockEthereumDebtCacheContract
        case mockChainConfig.optimism.debtCacheAddress:
          return mockOptimismDebtCacheContract
        case mockChainConfig.ethereum.synthetixDebtShareAddress:
        case mockChainConfig.optimism.synthetixDebtShareAddress:
          return mockSynthetixDebtShareContract
        default:
          break
      }
    },
  },
}))

let oldEnv: NodeJS.ProcessEnv

beforeAll(() => {
  oldEnv = JSON.parse(JSON.stringify(process.env))
  process.env.ETHEREUM_RPC_URL = mockChainConfig.ethereum.rpcUrl
  process.env.ETHEREUM_ADDRESS_PROVIDER_CONTRACT_ADDRESS =
    mockChainConfig.ethereum.addressProviderContractAddress
  process.env.OPTIMISM_RPC_URL = mockChainConfig.optimism.rpcUrl
  process.env.OPTIMISM_ADDRESS_PROVIDER_CONTRACT_ADDRESS =
    mockChainConfig.optimism.addressProviderContractAddress
  process.env.CACHE_ENABLED = 'false'
})

afterAll(() => {
  process.env = oldEnv
})

describe('synthetix-debt-pool', () => {
  let server: http.Server
  let req: SuperTest<Test>

  beforeAll(async () => {
    server = await startServer()
    req = request(`localhost:${(server.address() as AddressInfo).port}`)
  })

  afterAll((done) => {
    server.close(done)
  })

  describe('when making a request to fetch the current debt', () => {
    it('successfully fetches the current debt size of the synthetix debt cache across all chains if chainSources is missing', async () => {
      const request = {
        id: 1,
        data: {},
      }
      const response = await req
        .post('/')
        .send(request)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)

      expect(response.body).toMatchSnapshot()
    })

    it('successfully fetches the current debt size of the synthetix debt cache across all chains if chainSources is empty', async () => {
      const request = {
        id: 1,
        data: {
          chainSources: null,
        },
      }
      const response = await req
        .post('/')
        .send(request)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)

      expect(response.body).toMatchSnapshot()
    })

    it('successfully fetches the current debt size of the synthetix debt cache for only one chain', async () => {
      const request = {
        id: 1,
        data: {
          chainSources: ['ethereum'],
        },
      }
      const response = await req
        .post('/')
        .send(request)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)

      expect(response.body).toMatchSnapshot()
    })
  })

  describe('errors', () => {
    it('throws an error if the request contains a source without a chain configuration', async () => {
      const request = {
        id: 1,
        data: {
          chainSources: ['ethereum-fake'],
        },
      }
      const response = await req
        .post('/')
        .send(request)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(500)

      expect(response.body).toMatchSnapshot()
    })
  })
})
