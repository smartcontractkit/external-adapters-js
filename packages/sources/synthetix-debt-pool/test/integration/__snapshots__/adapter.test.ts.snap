// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`synthetix-debt-pool errors throws an error if the request contains a source without a chain configuration 1`] = `
Object {
  "error": Object {
    "feedID": "{\\"data\\":{\\"chainSources\\":[\\"ethereum-fake\\"],\\"resultPath\\":\\"debt\\"}}",
    "message": "Missing configuration for chain: ETHEREUM-FAKE",
    "name": "AdapterError",
  },
  "jobRunID": 1,
  "status": "errored",
  "statusCode": 500,
}
`;

exports[`synthetix-debt-pool when making a request to fetch the current debt successfully fetches the current debt size of the synthetix debt cache across all chains if chainSources is empty 1`] = `
Object {
  "data": Object {
    "result": "0x000000000162e5f31ae1af41dd78e49800000000010ce52dd3af66081c5b68ee",
  },
  "jobRunID": 1,
  "result": "0x000000000162e5f31ae1af41dd78e49800000000010ce52dd3af66081c5b68ee",
  "statusCode": 200,
}
`;

exports[`synthetix-debt-pool when making a request to fetch the current debt successfully fetches the current debt size of the synthetix debt cache across all chains if chainSources is missing 1`] = `
Object {
  "data": Object {
    "result": "0x000000000162e5f31ae1af41dd78e49800000000010ce52dd3af66081c5b68ee",
  },
  "jobRunID": 1,
  "result": "0x000000000162e5f31ae1af41dd78e49800000000010ce52dd3af66081c5b68ee",
  "statusCode": 200,
}
`;

exports[`synthetix-debt-pool when making a request to fetch the current debt successfully fetches the current debt size of the synthetix debt cache for only one chain 1`] = `
Object {
  "data": Object {
    "result": "0x0000000000b172f98d70d7a0eebc724c0000000000e2be35af6b47384f7b98ca",
  },
  "jobRunID": 1,
  "result": "0x0000000000b172f98d70d7a0eebc724c0000000000e2be35af6b47384f7b98ca",
  "statusCode": 200,
}
`;
